SSA/Projects/InstCombine/AliveHandwrittenExamples.lean

```diff
@@ -1,40 +1,81 @@
-
 import SSA.Projects.InstCombine.LLVM.EDSL
 import SSA.Projects.InstCombine.AliveStatements
 import SSA.Projects.InstCombine.Refinement
 import SSA.Projects.InstCombine.Tactic
+import SSA.Projects.InstCombine.Base
+import SSA.Core.ErasedContext
 
 open MLIR AST
 open Std (BitVec)
-open Ctxt (Var)
+open Ctxt (Var DerivedCtxt)
+open InstCombine (MOp)
 
-namespace AliveAutoGenerated
+namespace AliveHandwritten
 set_option pp.proofs false
 set_option pp.proofs.withType false
 
 
 /-
 Name: SimplifyDivRemOfSelect
-
+precondition: true
 %sel = select %c, %Y, 0
 %r = udiv %X, %sel
   =>
 %r = udiv %X, %Y
 
 -/
+def alive_DivRemOfSelect_src (w : Nat) :=
+  [alive_icom (w)| {
+  ^bb0(%c: i1, %y : _, %x : _):
+    %c0 = "llvm.mlir.constant" () { value = 0 : _ } :() -> (_)
+    %v1 = "llvm.select" (%c,%y, %c0) : (i1, _, _) -> (_)
+    %v2 = "llvm.udiv"(%x, %v1) : (_, _) -> (_)
+    "llvm.return" (%v2) : (_) -> ()
+  }]
+
+def alive_DivRemOfSelect_tgt (w : Nat) :=
+  [alive_icom (w)| {
+  ^bb0(%c: i1, %y : _, %x : _):
+    %v1 = "llvm.udiv" (%x,%y) : (_, _) -> (_)
+    "llvm.return" (%v1) : (_) -> ()
+  }]
+
+@[simp]
+theorem BitVec.ofNat_toNat_zero :
+BitVec.toNat (BitVec.ofInt w 0) = 0 :=
+  by simp[BitVec.toNat, BitVec.ofInt, BitVec.toFin, BitVec.ofNat, OfNat.ofNat]
+
+theorem alive_DivRemOfSelect (w : Nat) :
+    alive_DivRemOfSelect_src w ⊑ alive_DivRemOfSelect_tgt w := by
+  unfold alive_DivRemOfSelect_src alive_DivRemOfSelect_tgt
+  intros Γv
+  simp_peephole at Γv
+  simp (config := {decide := false}) only [OpDenote.denote,
+    InstCombine.Op.denote, HVector.toPair, HVector.toTriple, pairMapM, BitVec.Refinement,
+    bind, Option.bind, pure, DerivedCtxt.ofCtxt, DerivedCtxt.snoc,
+    Ctxt.snoc, ConcreteOrMVar.instantiate, Vector.get, HVector.toSingle,
+    LLVM.and?, LLVM.or?, LLVM.xor?, LLVM.add?, LLVM.sub?,
+    LLVM.mul?, LLVM.udiv?, LLVM.sdiv?, LLVM.urem?, LLVM.srem?,
+    LLVM.sshr, LLVM.lshr?, LLVM.ashr?, LLVM.shl?, LLVM.select?,
+    LLVM.const?, LLVM.icmp?,
+    HVector.toTuple, List.nthLe, bitvec_minus_one]
```

Comment:
Can these simp-lemmas be passed to `simp_peephole`? Do we want them to?

Reply:
If we like, they should probably be part of `simp_alive_peephole`, but as seen, this is fairly brittle :(